grakn-test-tools/src/main/java/ai/grakn/engine/rpc/SynchronousObserver.java

```diff
@@ -0,0 +1,190 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.engine.rpc;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Preconditions;
+import io.grpc.stub.StreamObserver;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+/**
+ * Wrapper for synchronous bidirectional streaming communication - i.e. when there is a stream of {@link Request}s and
+ * a stream of {@link Response}s.
+ *
+ * A request is sent with the {@link #send(Request)}} method, and you can block for a response with the
+ * {@link #receive()} method.
+ *
+ * <pre>
+ * {@code
+ *
+ *     try (SynchronousObserver<TxRequest, TxResponse> tx = SynchronousObserver.create(stub::tx) {
+ *         tx.send(openMessage);
+ *         TxResponse doneMessage = tx.receive().elem();
+ *         tx.send(commitMessage);
+ *         Throwable validationError = tx.receive.throwable();
+ *     }
+ * }
+ * </pre>
+ *
+ * @author Felix Chapman
+ *
+ * @param <Request> The type of requests being received
+ * @param <Response> The type of responses being sent
+ */
+public class SynchronousObserver<Request, Response> implements AutoCloseable {
+
+    private final StreamObserver<Request> requests;
+    private final QueueingObserver<Response> responses;
+
+    private SynchronousObserver(StreamObserver<Request> requests, QueueingObserver<Response> responses) {
+        this.requests = requests;
+        this.responses = responses;
+    }
+
+    /**
+     * Create a {@link SynchronousObserver} using a method that accepts a {@link StreamObserver} and returns
+     * a {@link StreamObserver}.
+     *
+     * <p>
+     * This looks super-weird because it is super-weird. The reason is that the gRPC-generated methods on client-stubs
+     * for bidirectional streaming calls look like this:
+     * </p>
+     *
+     * <p>
+     *     {@code StreamObserver<TxRequest> tx(StreamObserver<TxResponse> responseObserver)}
+     * </p>
+     *
+     * <p>
+     *     So, we cannot get at the {@link Request} observer until we provide a {@link Response} observer.
+     *     Unfortunately the latter is created within this method below - so we must pass in a method reference:
+     * </p>
+     *
+     * {@code SynchronousObserver.create(stub::tx)}
+     */
+    public static <Request, Response> SynchronousObserver<Request, Response> create(
+            Function<StreamObserver<Response>, StreamObserver<Request>> createRequestObserver
+    ) {
+        QueueingObserver<Response> responses = new QueueingObserver<>();
```

Comment:
Maybe state that you are only wrapping the responses into a blocking approach. I.e. you can still send requests asynchronously.